pkg/health/checker/checker.go

```diff
@@ -361,224 +180,347 @@ func watchConsulHealth(
 	}
-	host, _, err := net.SplitHostPort(u.Host)
+}
+
+func kvpToResult(kv api.KVPair) (*health.Result, error) {
+	res := &health.Result{}
+	err := json.Unmarshal(kv.Value, &res)
 	if err != nil {
-		return "", util.Errorf("error parsing host:port '%s'", u.Host)
+		return nil, util.Errorf("Could not unmarshal health at %s: %v", kv.Key, err)
 	}
-	return types.NodeName(host), nil
+	return res, nil
 }
 
-func healthResultsCopy(healthResults map[types.NodeName]health.Result) map[types.NodeName]health.Result {
-	out := make(map[types.NodeName]health.Result)
-	for nodeID, healthResult := range healthResults {
-		out[nodeID] = healthResult
+// Maps a list of KV Pairs into a slice of health.Results
+// Halts and returns upon encountering an error
+func kvpsToResult(kvs api.KVPairs) ([]*health.Result, error) {
+	result := make([]*health.Result, len(kvs))
+	for i, kv := range kvs {
+		res, err := kvpToResult(*kv)
+		if err != nil {
+			return nil, err
+		}
+		result[i] = res
 	}
-	return out
+
+	return result, nil
 }
 
-func (h shadowTrafficHealthChecker) WatchService(
+type HealthClient interface {
+	HealthCheckEndpoints(ctx context.Context, req *hclient.HealthEndpointsRequest) (map[string]health.HealthState, error)
+	HealthMonitor(ctx context.Context, req *hclient.HealthRequest, respCh chan *hclient.HealthResponse) error
+}
+
+type healthChecker struct {
+	healthClient HealthClient
+	serviceNodes func(serviceID string) ([]types.NodeName, error)
+}
+
+func NewHealthChecker(
+	hClient HealthClient,
+	serviceNodes func(serviceID string) ([]types.NodeName, error),
+) HealthChecker {
+	return healthChecker{
+		healthClient: hClient,
+		serviceNodes: serviceNodes,
+	}
+}
+
+func (h healthChecker) WatchPodOnNode(
 	ctx context.Context,
-	serviceID string,
-	nodeIDs []types.NodeName,
-	nodeIDsCh <-chan []types.NodeName,
-	resultCh chan<- map[types.NodeName]health.Result,
-	errCh chan<- error,
-	watchDelay time.Duration,
-	useHealthService bool,
-	useOnlyHealthService bool,
+	nodeID types.NodeName,
+	podID types.PodID,
 	status manifest.StatusStanza,
-) {
-	// nodes are considered always healthy when status port is not set
-	// when always healthy, WatchService ONLY sends the healthy status repeatedly on a timer set to delay
-	alwaysHealthy := status.Port == 0
-	if h.useHealthService || useHealthService || useOnlyHealthService || h.useOnlyHealthService {
-		respChan := make(chan *hclient.HealthResponse, len(nodeIDs))
-		go func() {
-			oldResultsDelay := time.Second * 5
-			healthResults := make(map[types.NodeName]health.Result)
-			if alwaysHealthy {
-				// reduce the delay to send old results since
-				// when the nodes are always healthy, these
-				// are the only messages sent
-				oldResultsDelay = time.Second * 2
-				for _, nodeID := range nodeIDs {
-					healthResults[nodeID] = health.Result{
-						ID:      types.PodID(serviceID),
-						Node:    nodeID,
-						Service: serviceID,
-						Status:  health.Passing,
+) (chan health.Result, chan error) {
```

Comment:
Does this code need a special case for always healthy?

Reply:
Yes, probably.